Merchant product overview

atiftonat opened the pull request.

BenTalese committed:
Share launch settings

BenTalese committed:
Add models

BenTalese committed:
Add services

BenTalese committed:
Add stock overview

BenTalese committed:
Update main layout

atiftonat committed:
Add card component

atiftonat committed:
Add youtube layout

atiftonat committed:
add product search page

atiftonat committed:
Add merchant & product store

atiftonat committed:
fix interpolated string in CreateProductPresenter.present_product_already_exists_async

atiftonat committed:
add save product action to product store

atiftonat committed:
add icon click event to card component

atiftonat committed:
refactor search bar & wire up icon click event

atiftonat committed:
Add dropdown component

atiftonat committed:
Add select component

atiftonat committed:
rename product model for consistency

atiftonat committed:
add get all dapi products to service

atiftonat committed:
Create interface and implement for sort options, folder likely to change

atiftonat committed:
implement product service & sort by options in product store

atiftonat committed:
implement sort options in product search, & experimental work

atiftonat committed:
fix namespaces & file name for sort options

atiftonat committed:
rename ProductSearchSortOption

atiftonat committed:
add ProductSearchFilterOptions

atiftonat committed:
refactor product store and add toggle actions + misc string fix

atiftonat committed:
add slot to select component to add icon on multiple usage for UX + in progress work for iconName prop

atiftonat committed:
add filters to product search + refatoring

atiftonat committed:
add & implement quasar loader plugin

atiftonat committed:
refactor card component w/ default props & misc select component changes

atiftonat committed:
Move initial app wide loads to app component

atiftonat committed:
refactor sort by options

atiftonat committed:
remove legacy sort by options

atiftonat committed:
create helpers for logic that does not fit within the pinia state

atiftonat committed:
refactor store by 1) composition api 2) logical concerns

atiftonat committed:
refactor card, select components

atiftonat committed:
refactor product search page

atiftonat committed:
add display if no products are found via search

atiftonat committed:
add img for display when no products are found via search

atiftonat committed:
suffix async method names, re-name scrapedProductOffer merchant to merchant_name , add specials filter & misc/bug fixes

atiftonat committed:
move sorting functionality to scraped product offer logic helper class

atiftonat committed:
add bandaid fix for image UTF error

atiftonat committed:
Select Component: Improve readability off multiple option logic and document component props

atiftonat committed:
remove quasar stock components/files

atiftonat committed:
Change sort option to set default without sting hardcoding

atiftonat committed:
misc refactoring

atiftonat committed:
Add vscode to .gitignore as on QuasarApp branch prior to merge

BenTalese commented on .vscode/settings.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "files.exclude": {
+        "**/__pycache__": true,
```
perhaps we want to update the gitignore file so that we ignore everything except the tasks and launch files?

atiftonat committed:
move inline styling to scoped style section

atiftonat commented on framework/web_app/quasar.config.ts:
```diff
@@ -133,7 +144,7 @@ export default configure((/* ctx */) => {
     // https://v2.quasar.dev/quasar-cli-vite/developing-ssr/configuring-ssr
     ssr: {
-      prodPort: 3000, // The default port that the production server should use
+      prodPort: 5174, // The default port that the production server should use
```
Not sure why the prodPort changed, does this impact you @BenTalese ?

atiftonat replied:
dev server also changed, auto-generated changes

atiftonat replied:
@BenTalese  plz confirm if fine

atiftonat reviewed (commented):
review #1

atiftonat commented on framework/dora_api/view_models/product_view_model.py:
```diff
@@ -25,7 +25,7 @@ class ProductViewModel:
     return ProductViewModel(
         brand = product.brand,
-        image = product.image.decode('utf-8'),
+        image = product.image.decode('utf-8', 'ignore'),
```
document why/what

atiftonat commented on framework/web_app/src/App.vue:
```diff
@@ -0,0 +1,18 @@
+
+  const productStore = useProductStore();
+
+  //TODO: If initial load up fails.
```
this is a separate issue, add to the excel file and resolve

atiftonat committed:
Remove redundant default logic for card component icon, the prop was not even optional

BenTalese commented on framework/web_app/src/services/api/ProductApiService.ts:
```diff
@@ -0,0 +1,44 @@
+        this.mapiHttpClient = new AxiosHttpClient(5172);
+    }
+
+    // TODO: Need to decide how to deal with delete product
```
After thinking over everything including our past discussions, i reckon once a product is created, that's it, it's in the database. If we really need to, we can include a deletion button somewhere, maybe on the product itself under the "my products" page... but that would require confirmation + what benefit does the end user get out of it? I think we're the only ones likely to think about database size and redundant information. The average user sees it all as magic voodoo. We could later on include a "delete all data" button...I reckon that would be the most useful to someone if they wanted to start fresh. If you agree we can add that to the spreadsheet.

So with that in mind, we can include a flag "is archived" on the product, which is flagged as true when a product is "unfavourited" or "unsaved". An archived product I guess we wouldn't scrape for deals on, and would be hidden from everywhere except the product search page so you can re-save it. I think it'd be a nicer end-user experience (as you previously said) to have previously scrapped deals stored in case they want to use that product again in future. Also what if someone accidentally unsaves a product? They then lose X amount of saved deals? Seems bad.

Thoughts?

atiftonat replied:
1. Agree that we just have some sort of active or saved flag and don't delete. Don't think we need to include a genuine delete for MVP. It can always be a feature request in the future for some reason.
2. Also think it is a good idea to filter out additional deals or logic for anything not active.
3. The user accidentally unsaving atm i'm not sure is a problem unless our UI is shit & you can randomly unsave products.
4. Currently we don't have the ability to unsave anyways. So i think we deal with this when we add support for the use case. **Plz resolve if happy with all.**

atiftonat replied:
Removed the comment & added a note to the dump.

BenTalese commented on framework/web_app/src/services/files/ImageService.ts:
```diff
@@ -0,0 +1,36 @@
+export default class ImageService {
+    // encodeToBase64 = (imageFile: File): Promise<string> =>
+    //     new Promise((resolve, reject) => {
```
we may want to include this commented out function if required for saving the image files

BenTalese replied:
wait no, i remember now...was never needed because the offer's original image is never touched, so you don't need to change it back

sooooo.... we could keep this encode function if we think we'd need it for some reason? hard to say but currently there's no use for it

plus we can delete the example below

atiftonat replied:
Inclined to leave comments & examples for now because saving images or retrieving saved images is not working atm, that's why I had that ignore.

Will address img's when doing the my products page.

atiftonat replied:
Resolve if happy with this

BenTalese commented on framework/web_app/src/stores/MerchantStore.ts:
```diff
@@ -0,0 +1,13 @@
+
+export const useMerchantStore = defineStore('merchant', {
+    state: () => ({
+
```
i vote move :smile:

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,160 @@
+
+// TODO:
+// New Features:
+// - Create a relevance algorithm & set as default sortBy
```
all good improvements it looks like, happy to explore either git issues or keep to excel sheets, but if we move to git i'd go through and add everything into git rather than have stuff spread out everywhere

atiftonat replied:
Fixed the issue with stockcode coming as numbers out of MAPI & added rest to the dump

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,163 @@
+
+// TODO:
+// Issues:
+// - Scraped offers return merchant stock code as a number BUT get products returns it as a string
```
probably want a string, no? i think stockcode could be anything

atiftonat replied:
is a string now

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,160 @@
+
+        shallowOffersCopy = shallowOffersCopy.filter(off =>
+            (offerFilters.showOnlyAvailable ? off.is_available === true : true) &&
+            (offerFilters.showOnlyFavourites ? isOfferFavourited(off, products.value) : true) &&
```
agreed, i don't think we need to filter to favourited

BenTalese replied:
i could see excluding them, but i doubt people would have that many favourites

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,160 @@
+        shallowOffersCopy = shallowOffersCopy.filter(off =>
+            (offerFilters.showOnlyAvailable ? off.is_available === true : true) &&
+            (offerFilters.showOnlyFavourites ? isOfferFavourited(off, products.value) : true) &&
+            (offerFilters.showOnlySpecials ?  off.price_now < off.price_was : true) &&
```
i like your thinking, keep the UI as simple and as easy to read as possible, makes our lives easier

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,160 @@
+            (offerFilters.showOnlyFavourites ? isOfferFavourited(off, products.value) : true) &&
+            (offerFilters.showOnlySpecials ?  off.price_now < off.price_was : true) &&
+            merchantNames.value?.includes(off.merchant_name) &&
+            // Don't display products that don't have a current price
```
agree with your thoughts, hiding would be bad UX, also we can't assume the product is unavailable or old, so we shouldn't hide it based on that

i reckon we can display "Price Unavailable" or "N/A" or something... hiding the product is bad and i don't think we should misrepresent pricing information by merging the two together

the other thing to consider is products will be marked as unavailable from the merchant...i think you can still get pricing information from unavailable products, so that may be a source for this problem with incomplete pricing information... so we should ensure we deal with that accordingly

btw how are you showing unavailable products? greyed out?

atiftonat replied:
1. went with price unavailable
2. the image has a caption when unavailable saying 'out of stock'

atiftonat replied:
if you're happy with the above plz resolve

BenTalese commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,160 @@
+        productApiService
+            .searchByTermAsync(query)
+            .then((offers) => productOffers.value = offers)
+            .catch(() => {})
```
shouldn't need to catch anywhere in the front end

the catches are in the api services and...i've yet to do anything with them lol, they just rethrow for now but we should log or display errors from there

we may want a global error modal or something? yeah even a banner

BenTalese reviewed (commented):
Have fun :stuck_out_tongue:

atiftonat committed:
Add documentation to card component props & emits

atiftonat committed:
Document reason for product view model's image decoding errors being ignored

atiftonat committed:
Add custom errors that are not built-in on javascript

atiftonat committed:
Create merchant store & utilise in the app file & on the product search stores filter

atiftonat committed:
Add helper method to retrieve the nameof a an instance member

atiftonat committed:
Refactor the product store (remove merchantApiService which is now in the merchantStore, misc renames & documentation, create generic toggle action for the product search filters)

atiftonat committed:
add message option to not supported error

atiftonat committed:
fix casing & naming on offerSortByOptions

atiftonat committed:
On the product search page: utilise the generic filter toggle action & misc renames

atiftonat committed:
On the product search page: Add a watch once functionality to initalise the store filter once merchants are loaded

BenTalese commented on .vscode/settings.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "files.exclude": {
```
Wrong file but reminder to only show filters for stores returned by product search

Alternatively, add endpoint to merchant API to get turned on merchants

atiftonat commented on framework/web_app/src/components/DropdownComponent.vue:
```diff
@@ -0,0 +1,40 @@
+        no-wrap
+        :persistent="false"
+        size="md"
+        :dense="true"
```
component was not used (select is being used) & deleted

atiftonat commented on framework/web_app/src/components/DropdownComponent.vue:
```diff
@@ -0,0 +1,40 @@
+        :stretch="false"
+        type="button"
+        transition-duration ="300"
+        transition-hide="fade"
```
component does not exist anymore, select is being used

atiftonat commented on framework/web_app/src/components/DropdownComponent.vue:
```diff
@@ -0,0 +1,40 @@
+<template>
+    <q-btn-dropdown
+        :auto-close="true"
```
component does not exist anymore, select is being used

atiftonat committed:
Make store state readonly, add scrapedOffer logic & remove filter logic that removes offers without a current price

atiftonat committed:
Change custom errors to utilise the object's prototype for it's name property

atiftonat committed:
Add back ignore hack that was accidentally removed

atiftonat committed:
Misc changes (remove comments, add comments, change casing & naming of OfferSortByOptions)

atiftonat committed:
rename MAPI router to be product router

atiftonat committed:
add get merchants to MAPI & replace the usage of DAPI for getting Merchants in the web app

atiftonat committed:
Move initial app wide fetches to a beforeMount lifecycle hook

atiftonat committed:
Remove redundant ID property from web app model

atiftonat committed:
Utilise the base app scss file for shared styles & variables

atiftonat committed:
For best performance move functions defined within the template into the script & misc renames

atiftonat committed:
Change card merchant display from text quasar chips to logos

atiftonat committed:
Add calculated values to MAPI & add sanitisation to simplify frontend

atiftonat committed:
Add 'SAVE $$$' chip to cards, hide price was if it does not exist & show 'Price Unavailable' if the current price does not exist

atiftonat committed:
Prefix interfaces with 'I'

atiftonat committed:
Add price per cup to offers

atiftonat committed:
Initialise stores filter in merchant store. Tried many approaches, this is currently the most consistent solution, other solutions such as onBeforeMount hooks within ProductSearch page results in race conditions occasionally which result in the filter's options not being selected as the default state

atiftonat committed:
Add price_per_cup to web app ScrapedProductOffer model

atiftonat committed:
remove router on MAPI that was superseded with more granular routers

atiftonat committed:
remove redundant onBeforeMount hook

atiftonat committed:
remove no longer relevant comment

atiftonat committed:
add caption for unavailable offers,
make stockcode come out as string from MAPI, improve price unavailable display,
add future work to ticket dump

atiftonat committed:
delete currently unused layout example

atiftonat commented on framework/web_app/src/layouts/YoutubeLayout.vue:
```diff
@@ -0,0 +1,236 @@
+<template>
+  <q-layout view="hHh lpR fFf" class="bg-grey-1">
+    <q-header elevated class="bg-white text-grey-8 q-py-xs" height-hint="58">
```
deleting, can get again from quasar website if we need

atiftonat commented on framework/dora_api/view_models/product_view_model.py:
```diff
@@ -25,7 +25,7 @@ class ProductViewModel:
     return ProductViewModel(
         brand = product.brand,
-        image = product.image.decode('utf-8'),
+        image = product.image.decode('utf-8', 'ignore'),
```
retrieving saved images throws erros, this is a temporary hack since this page doesnt care.
Will address when doing my products page

atiftonat commented on framework/merchant_api/domain/entities/scraped_product_offer.py:
```diff
@@ -19,7 +19,7 @@ class ScrapedProductOffer:
     image_uri: str
     is_available: bool
-    merchant: str # TODO: hmm...str? or id of merchant? would have to use GetMerchants or create if not found
+    merchant_name: str # TODO: hmm...str? or id of merchant? would have to use GetMerchants or create if not found
```
ye makes sense, removed

atiftonat commented on framework/web_app/src/App.vue:
```diff
@@ -3,5 +3,16 @@
+
+  const productStore = useProductStore();
+
+  //TODO: If initial load up fails.
```
added offline detection scenario to the dump & removed comment

atiftonat commented on framework/web_app/src/models/Product.ts:
```diff
@@ -0,0 +1,16 @@
+    web_url: string
+}
+
+// This is the view model with logic, callbacks, etc... maybe?
```
deleted comment, what we are doing currently seems fine

atiftonat commented on framework/web_app/src/models/StockItem.ts:
```diff
@@ -0,0 +1,20 @@
+/*
+Notes for later:
```
Yeah i think how we are using types is fine & meets our current requirements. Can always look into classes if we need more complexity/logic in the future.

atiftonat commented on framework/web_app/src/helpers/OfferSortByOptions.ts:
```diff
@@ -0,0 +1,36 @@
+        Apply: sortOffersByFullNameDesc
+    },
+    {
+        Description: 'Price (low to high)',
```
in bracket text has been capitalised for all

atiftonat commented on framework/web_app/src/helpers/ScrapedProductOfferLogic.ts:
```diff
@@ -0,0 +1,80 @@
+/**
+ * E.g., Coles offers don't always contain the brand in the name.
+ */
+export function getOfferFullName(brand: string | null, name: string): string {
```
this & some other logic has been moved to MAPI

atiftonat commented on framework/web_app/src/helpers/ScrapedProductOfferLogic.ts:
```diff
@@ -0,0 +1,80 @@
+/**
+ * E.g., Coles donuts have a null value and a '6 pack' unit.
+ */
+export const getOfferSize = (value: number, unit: string): string =>
```
this & some other logic has been moved to MAPI.
For this i added another property in MAPI that is the size val & unit together as a string

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+//#region Search
+
+const searchTerm = ref('');
+const previousSearchTerm = ref<undefined | string>();
```
think const is fine, the reference itself is never changing

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+const search = (): Promise<string> =>
+    productStore.searchByTermAsync({
+        search_term: searchTerm.value,
+        start_page: currentPage
```
I've been using the dump in excel.
As far as pagination goes i think we add it when it becomes a need, e.g., if we add more search results, more merchants.

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+    Woolworths: 'green-14'
+};
+
+const getMerchantColour = (merchantName: string): string =>
```
This got superseded by something for logos instead.

There is UI or logic  i'll have to make reusable/shared once i do the my products page because it is similar to this page.

For now, i think it is fine.

atiftonat commented on framework/web_app/src/helpers/ScrapedProductOfferLogic.ts:
```diff
@@ -0,0 +1,80 @@
+
+export const sortOffersBySpecialBestToWorst = (offers: ScrapedProductOffer[]): ScrapedProductOffer[] =>
+    offers.sort((offerA, offerB) => {
+        const _OfferA = (offerA.price_was ?? 0) - (offerA.price_now ?? 0);
```
i decided to just set it to 0 in MAPI instead of nulls, not sure if ideal but it did remove a bunch of checks for me (in here + UI)

atiftonat commented on framework/web_app/src/stores/ProductStore.ts:
```diff
@@ -0,0 +1,163 @@
+
+    //#endregion Filters
+
+    //#region Merchants
```
has been moved

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+
+        <q-btn
+            type="button"
+            @click="productStore.toggleShowOnlyFavourites"
```
removed

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+
+        <q-btn
+            type="button"
+            @click="productStore.toggleShowOnlySpecials"
```
we discussed in person & decide it was fine because Woolies does the same

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+                    <q-card-section class="q-py-none" style="flex: 1;">
+                        <div class="column full-height no-wrap justify-between">
+                            <div class="text-body2 text-weight-regular q-pb-sm">
+                                {{ `${getOfferFullName(offer.brand, offer.name)} | ${getOfferSize(offer.size_value, offer.size_unit)}` }}
```
both are now handled by MAPI

atiftonat commented on framework/web_app/src/pages/ProductSearch.vue:
```diff
@@ -0,0 +1,197 @@
+                            </div>
+                            <div>
+                                <div class="text-body2 text-weight-medium">
+                                    {{ `$${offer.price_now?.toFixed(2).toString()}` }}
```
is now handled by MAPI

atiftonat committed:
remove comment & newline

atiftonat committed:
Delete framework/web_app/.vscode/extensions.json

added to gitignore

atiftonat committed:
remove settings file from cache, impeding gitignore

atiftonat committed:
remove settings file from cache, impeding gitignore

atiftonat committed:
Delete .vscode/settings.json

added to gitignore

atiftonat committed:
remove file from cache that should be ignored

atiftonat committed:
Delete framework/web_app_old/.vscode/extensions.json

added to git ignore

atiftonat committed:
Delete framework/web_app/.vscode/settings.json

added to gitignore

atiftonat committed:
Update gitignore

atiftonat committed:
Merge branch 'MerchantProductOverview' of https://github.com/BenTalese/DiscountDora into MerchantProductOverview

BenTalese commented on framework/web_app/src/helpers/OfferSortByOptions.ts:
```diff
@@ -0,0 +1,36 @@
+        sort: sortOffersByFullNameDesc
+    },
+    {
+        description: 'Price (Low to High)',
```
when i shop, the most essential thing to me is the unit price, see woolies has sort by "Unit Price (High to Low)" and "Unit Price (Low to High)"

these i reckon we should definitely do, although i can see how they'd be difficult as we'd have to do quite a bit of conversions

happy to make this a future feature on the excel sheet, thoughts?

atiftonat replied:
Can discuss, need to think about how to approach this. But yeah i think we push it out into a separate PR at this point so this stuff goes through.

BenTalese reviewed (commented):
Just a tad left

atiftonat committed:
abc properties for coles & woolies product offer entities

atiftonat committed:
Move name logic out of ScrapedProductOffer (currently issue specifically for Coles translation). Also move size out of post init, this never needed to be calculated.

atiftonat commented on framework/merchant_api/domain/entities/scraped_product_offer.py:
```diff
@@ -19,15 +19,26 @@ class ScrapedProductOffer:
 
+    # TODO LOOk if can do in method before translates
+    def __post_init__(self):
+        self.name = self.name if self.brand in self.name else f"{self.brand} {self.name}"
```
1. Moved self.name. 
Think it was only Coles with the issue. If Woolies has the issue too we can add the logic to their translator method.

2. self.price_difference is calculated so that should stay

3. Moved self.size
IDK why i was calculating it, that value is already being scraped so i am now using that

4. Should logic to sanitise price now & price was to 2 decimals places stay in the post init?
I imagine we want that for all merchants?

BenTalese replied:
yeah that all makes sense, happy to leave it as is now

atiftonat commented on framework/merchant_api/domain/entities/woolworths_product_offer.py:
```diff
@@ -13,3 +13,5 @@ class WoolworthsProductOffer(BaseModel, extra='allow'):
     InstoreWasPrice: float
     LargeImageFile: str # 'https://cdn0.woolworths.media/content/wowproductimages/large/114682.jpg'
     PackageSize: str # '333G'
+    CupString: str # '$0.80 / 100G'
```
Did this manually for now.
What is the extension name for future reference?

atiftonat commented on framework/web_app/src/components/ColesLogo.vue:
```diff
@@ -0,0 +1,53 @@
+<template>
+    <svg
+        :height="height"
```
They just do it differently on their websites.
Took the implementation from their logos on their websites.

atiftonat commented on framework/web_app/src/components/CardComponent.vue:
```diff
@@ -0,0 +1,116 @@
+
+        <slot name="footer"></slot>
+
+      </div>
```
Will need to look into this. I reinstalled prettier but doesnt seem to have done the trick.

atiftonat committed:
Format components with vue (tab width 4 & more)
